<a href="https://colab.research.google.com/github/AlienX77-cmd/Python-for-Business-and-Finance/blob/main/%20Part_2_Stats_Exercise_04.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Coding Exercises (Part 2)

Now, you will have the opportunity to practice what you have learned. <br>
__Follow the instructions__ and insert your code! 

The correct answer is provided below your coding cell. There you can check whether your code is correct.

If you need some help or if you want to check your code, you can also have a look at the __solutions__.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Exercise 4: Hypothesis Testing

You have a random sample (drawn from S&P 500) with 50 stocks/companies and their annual returns for the year 2018 (__sample size = 50__)

1. Import the sample. 

In [7]:
# run the cell
import numpy as np
import scipy.stats as stats
np.set_printoptions(precision=4, suppress= True)

In [8]:
# run the cell
sample = np.loadtxt("sample_2018.csv", delimiter = ",", usecols = 1)
sample

array([ -3.189 ,  -0.7402, -10.0045,  -0.6494,  -6.7991, -10.3929,
       -19.4511, -21.3489, -23.8865, -24.194 ,  19.9705,  -2.8071,
        -7.9205,   2.8771, -42.3216,   1.7465, -54.1992, -20.5551,
        -0.0258,  36.4959, -22.8224, -57.5208, -10.6338,  -9.1523,
         9.707 , -21.1643, -23.9187,   7.2802, -45.8564,  33.4487,
         0.0822,  -7.8966, -27.5657,   9.678 , -15.3846,  35.2119,
         7.1421,  17.2502,  73.3463,  35.6039,  18.6085,  27.3282,
        -0.8907,   4.4209, -27.6435, -18.8478,  10.0865,  22.4127,
        -6.1347, -20.2881])

2. Test on a __10% level of significance__, whether the (equally-weighted) mean return for the whole S&P 500 population for the year 2018 is __equal to 0%__ or not. Formulate H0 and Ha and calculate the __t-statistic__ and the __p-value__ of your test.

In [29]:
#H0: mean return == 0%
#Ha: mean return != 0%
H0 = 0
stats.ttest_1samp(sample, H0)

TtestResult(statistic=-1.1023818747645209, pvalue=0.27568087004548397, df=49)

(result: Ttest_1sampResult(statistic=-1.1023818747645209, pvalue=0.27568087004548397)

3. Can you __reject H0__ (and conclude that the true mean is significantly different from 0%)?

In [30]:
conf = 0.9

In [31]:
t_stat, p_value = stats.ttest_1samp(sample, H0)

In [32]:
t_stat, p_value

(-1.1023818747645209, 0.27568087004548397)

In [33]:
p_value < 1-conf

False

4. Test on a __5% level of significance__, whether the (equally-weighted) mean return for the whole S&P 500 population for the year 2018 is __equal to or greater than 5%__ (H0). Can you conclude that the mean return is lower than 5% (Ha)?

In [34]:
#run the cell!
H0 = 5

In [35]:
t_stat, p_value = stats.ttest_1samp(sample, H0)

In [36]:
t_stat, p_value

(-2.5414652019770783, 0.01425369761073307)

In [37]:
p_value = p_value / 2
p_value

0.007126848805366535

In [38]:
conf = 0.95
p_value < 1-conf

True

(result: pvalue = 0.007126848805366535, reject H0)

5. Import the full __S&P 500 population__ (annual returns) for the year 2018.

In [10]:
# run the cell
pop = np.loadtxt("SP500_2018.csv", delimiter = ",", usecols = 1)
pop

array([ -1.5081, -38.8238,  55.9434,  -7.6819,  -5.7802, -19.4511,
        69.7722,  24.5663,  -8.6681,  27.3282,  -4.0099,   1.7465,
        10.2739, -40.8947,  21.3679,  10.0865,   1.5767,  31.7636,
         2.4151, -18.8165, -35.1796,  -0.5262, -12.5744,  14.6018,
        -6.1347, -39.9953,  -6.2559, -17.9567, -21.6694,  -1.2695,
       -18.7139, -36.6588, -21.1864,  73.3463,  -7.9205, -53.2229,
         9.678 , -39.1338,  11.3268,  26.3837, -12.6581,  -3.1167,
        15.1415,   7.2388, -30.3198, -37.6125,  -2.06  ,  -8.3485,
       -27.6435, -12.7498, -38.4954,   6.7179, -26.0897,  -3.2173,
        -1.2884, -22.6624,  -1.6067,  -4.924 ,  17.9428,   7.2802,
       -17.378 ,   0.8818, -13.6163, -25.0183,   3.7046, -32.6794,
       -13.6877, -32.2377,  -7.8633, -13.3866,  -1.2608, -24.4262,
        18.362 , -16.8734,   5.0232,   1.9473,  39.8548, -32.4525,
       -13.9199, -30.3454, -43.7749, -27.2401, -20.2881, -12.5368,
       -22.7787,  -7.8966, -26.4237,  -2.6034, -26.6536,   3.6

6. Calculate the population´s __skew__ and __kurtosis__.

In [39]:
stats.skew(pop)

0.41505283064889537

In [40]:
stats.kurtosis(pop)

0.6618700837691662

7. Test for __skew__ on a __5% significance__ level. 

In [41]:
stats.skewtest(pop)

SkewtestResult(statistic=3.700813132575524, pvalue=0.00021490969981373073)

(results: SkewtestResult(statistic=3.700813132575524, pvalue=0.00021490969981373073))

8. Test for __kurtosis__ on a __5% significance__ level.

In [42]:
stats.kurtosistest(pop)

KurtosistestResult(statistic=2.5240612693439624, pvalue=0.011600770790237903)

(results: KurtosistestResult(statistic=2.5240612693439624, pvalue=0.011600770790237903))

9. Test for __Normality__ on a __5% significance__ level!

In [43]:
stats.normaltest(pop)

NormaltestResult(statistic=20.066903133645717, pvalue=4.390635151897616e-05)

(results: NormaltestResult(statistic=20.066903133645717, pvalue=4.390635151897616e-05))

# Well Done!

---------------------------------------------------------------------------------------------------------------------

# Solutions (Stop here if you want to code on your own!)

# ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

You have a random sample (drawn from S&P 500) with 50 stocks/companies and their annual returns for the year 2018 (__sample size = 50__)

1. Import the sample. 

In [11]:
# run the cell
import numpy as np
import scipy.stats as stats
np.set_printoptions(precision=4, suppress= True)

In [12]:
# run the cell
sample = np.loadtxt("sample_2018.csv", delimiter = ",", usecols = 1)
sample

array([ -3.189 ,  -0.7402, -10.0045,  -0.6494,  -6.7991, -10.3929,
       -19.4511, -21.3489, -23.8865, -24.194 ,  19.9705,  -2.8071,
        -7.9205,   2.8771, -42.3216,   1.7465, -54.1992, -20.5551,
        -0.0258,  36.4959, -22.8224, -57.5208, -10.6338,  -9.1523,
         9.707 , -21.1643, -23.9187,   7.2802, -45.8564,  33.4487,
         0.0822,  -7.8966, -27.5657,   9.678 , -15.3846,  35.2119,
         7.1421,  17.2502,  73.3463,  35.6039,  18.6085,  27.3282,
        -0.8907,   4.4209, -27.6435, -18.8478,  10.0865,  22.4127,
        -6.1347, -20.2881])

2. Test on a __10% level of significance__, whether the (equally-weighted) mean return for the whole S&P 500 population for the year 2018 is __equal to 0%__ or not. Formulate H0 and Ha and calculate the __t-statistic__ and the __p-value__ of your test.

__H0:__ mean return == 0% <br> 
__Ha:__ mean return != 0%

In [13]:
H0 = 0

In [14]:
stats.ttest_1samp(sample, H0)

TtestResult(statistic=-1.1023818747645209, pvalue=0.27568087004548397, df=49)

(result: Ttest_1sampResult(statistic=-1.1023818747645209, pvalue=0.27568087004548397)

3. Can you __reject H0__ (and conclude that the true mean is significantly different from 0%)?

In [15]:
conf = 0.9

In [16]:
t_stat, p_value = stats.ttest_1samp(sample, H0)

In [17]:
p_value < 1-conf

False

4. Test on a __5% level of significance__, whether the (equally-weighted) mean return for the whole S&P 500 population for the year 2018 is __equal to or greater than 5%__ (H0). Can you conclude that the mean return is lower than 5% (Ha)?

In [18]:
#run the cell!
H0 = 5

In [19]:
t_stat, p_value = stats.ttest_1samp(sample, H0)

In [20]:
t_stat

-2.5414652019770783

In [21]:
p_value = p_value / 2

In [22]:
p_value

0.007126848805366535

(result: pvalue = 0.007126848805366535, reject H0)

5. Import the full __S&P 500 population__ (annual returns) for the year 2018.

In [23]:
# run the cell
pop = np.loadtxt("SP500_2018.csv", delimiter = ",", usecols = 1)
pop

array([ -1.5081, -38.8238,  55.9434,  -7.6819,  -5.7802, -19.4511,
        69.7722,  24.5663,  -8.6681,  27.3282,  -4.0099,   1.7465,
        10.2739, -40.8947,  21.3679,  10.0865,   1.5767,  31.7636,
         2.4151, -18.8165, -35.1796,  -0.5262, -12.5744,  14.6018,
        -6.1347, -39.9953,  -6.2559, -17.9567, -21.6694,  -1.2695,
       -18.7139, -36.6588, -21.1864,  73.3463,  -7.9205, -53.2229,
         9.678 , -39.1338,  11.3268,  26.3837, -12.6581,  -3.1167,
        15.1415,   7.2388, -30.3198, -37.6125,  -2.06  ,  -8.3485,
       -27.6435, -12.7498, -38.4954,   6.7179, -26.0897,  -3.2173,
        -1.2884, -22.6624,  -1.6067,  -4.924 ,  17.9428,   7.2802,
       -17.378 ,   0.8818, -13.6163, -25.0183,   3.7046, -32.6794,
       -13.6877, -32.2377,  -7.8633, -13.3866,  -1.2608, -24.4262,
        18.362 , -16.8734,   5.0232,   1.9473,  39.8548, -32.4525,
       -13.9199, -30.3454, -43.7749, -27.2401, -20.2881, -12.5368,
       -22.7787,  -7.8966, -26.4237,  -2.6034, -26.6536,   3.6

6. Calculate the population´s __skew__ and __kurtosis__.

In [24]:
stats.skew(pop)

0.41505283064889537

In [25]:
stats.kurtosis(pop)

0.6618700837691662

7. Test for __skew__ on a __5% significance__ level. 

In [26]:
stats.skewtest(pop)

SkewtestResult(statistic=3.700813132575524, pvalue=0.00021490969981373073)

(results: SkewtestResult(statistic=3.700813132575524, pvalue=0.00021490969981373073))

8. Test for __kurtosis__ on a __5% significance__ level.

In [27]:
stats.kurtosistest(pop)

KurtosistestResult(statistic=2.5240612693439624, pvalue=0.011600770790237903)

(results: KurtosistestResult(statistic=2.5240612693439624, pvalue=0.011600770790237903))

9. Test for __Normality__ on a __5% significance__ level!

In [28]:
stats.normaltest(pop)

NormaltestResult(statistic=20.066903133645717, pvalue=4.390635151897616e-05)

(results: NormaltestResult(statistic=20.066903133645717, pvalue=4.390635151897616e-05))